File: sites/vpn/mfa/totp.ts
/**
 * RFC 6238 TOTP (SHA-1, 6 digits, 30s). Any local authenticator app.
 * No Google account.
 */
import { createHmac, timingSafeEqual } from "node:crypto";

const ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";

export function decodeBase32(secret: string): Buffer {
  const clean = secret.toUpperCase().replace(/=+$/g, "").replace(/\s+/g, "");
  let bits = "";
  for (const ch of clean) {
    const idx = ALPHABET.indexOf(ch);
    if (idx < 0) {
      throw new Error("invalid base32");
    }
    bits += idx.toString(2).padStart(5, "0");
  }
  const bytes: number[] = [];
  for (let i = 0; i + 8 <= bits.length; i += 8) {
    bytes.push(Number.parseInt(bits.slice(i, i + 8), 2));
  }
  return Buffer.from(bytes);
}

export function totpAt(secret: Buffer, atMs: number, periodSec = 30): string {
  const counter = Math.floor(atMs / 1000 / periodSec);
  const buf = Buffer.alloc(8);
  buf.writeUInt32BE(Math.floor(counter / 0x100000000), 0);
  buf.writeUInt32BE(counter >>> 0, 4);
  const hmac = createHmac("sha1", secret).update(buf).digest();
  const offset = hmac[hmac.length - 1]! & 0xf;
  const bin =
    ((hmac[offset]! & 0x7f) << 24) |
    (hmac[offset + 1]! << 16) |
    (hmac[offset + 2]! << 8) |
    hmac[offset + 3]!;
  return String(bin % 1_000_000).padStart(6, "0");
}

export function totpOk(secretB32: string, code: string, atMs = Date.now()): boolean {
  const secret = decodeBase32(secretB32);
  const guess = code.replace(/\s+/g, "");
  if (!/^\d{6}$/.test(guess)) {
    return false;
  }
  const window = [0, -1, 1];
  const want = Buffer.from(guess);
  for (const w of window) {
    const ts = atMs + w * 30_000;
    if (ts < 0) {
      continue;
    }
    const have = Buffer.from(totpAt(secret, ts));
    if (have.length === want.length && timingSafeEqual(have, want)) {
      return true;
    }
  }
  return false;
}

File: sites/vpn/mfa/totp.py
#!/usr/bin/env python3
"""RFC 6238 TOTP (SHA-1, 6 digits, 30s). Works with any local authenticator app.

No network. No Google account.

Run:  python sites/vpn/mfa/server.py
Test: python sites/vpn/mfa/tests/test_totp.py
"""

from __future__ import annotations

import base64
import hashlib
import hmac
import struct
import time

ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567"


def decode_base32(secret: str) -> bytes:
    """Decode a base32 TOTP secret. Raises ValueError on bad input."""
    clean = (
        secret.upper()
        .replace("=", "")
        .replace(" ", "")
        .replace("\t", "")
        .replace("\n", "")
    )
    clean = clean.replace("\r", "")
    if not clean:
        raise ValueError("invalid base32")
    for ch in clean:
        if ch not in ALPHABET:
            raise ValueError("invalid base32")
    padded = clean + "=" * ((8 - len(clean) % 8) % 8)
    try:
        return base64.b32decode(padded, casefold=True)
    except Exception as exc:
        raise ValueError("invalid base32") from exc


def totp_at(secret: bytes, at_ms: float, period_sec: int = 30) -> str:
    """Return the 6-digit code for secret at at_ms (epoch millis)."""
    counter = int(at_ms // 1000 // period_sec)
    msg = struct.pack(">Q", counter)
    digest = hmac.new(secret, msg, hashlib.sha1).digest()
    offset = digest[-1] & 0x0F
    code = (
        ((digest[offset] & 0x7F) << 24)
        | (digest[offset + 1] << 16)
        | (digest[offset + 2] << 8)
        | digest[offset + 3]
    )
    return str(code % 1_000_000).zfill(6)


def totp_ok(secret_b32: str, code: str, at_ms: float | None = None) -> bool:
    """Check code against a +-1 step window. Constant-time compare."""
    try:
        secret = decode_base32(secret_b32)
    except ValueError:
        return False
    guess = code.replace(" ", "").replace("\t", "").replace("\n", "")
    if len(guess) != 6 or not guess.isdigit():
        return False
    now_ms = float(at_ms) if at_ms is not None else time.time() * 1000
    for window in (0, -1, 1):
        shifted = now_ms + window * 30_000
        if shifted < 0:
            continue
        candidate = totp_at(secret, shifted)
        if hmac.compare_digest(candidate, guess):
            return True
    return False
